subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainDependencyToComponentIdResolver.java

```diff
@@ -46,18 +48,25 @@ public void add(ModuleComponentRepository repository) {
 
     public void resolve(DependencyMetadata dependency, ModuleIdentifier targetModuleId, BuildableComponentIdResolveResult result) {
         ModuleVersionSelector requested = dependency.getRequested();
-        VersionSelector versionSelector = versionSelectorScheme.parseSelector(requested.getVersion());
-        if (versionSelector.isDynamic()) {
-            dynamicRevisionResolver.resolve(dependency, versionSelector, result);
+        VersionConstraintInternal constraint = (VersionConstraintInternal) requested.getVersionConstraint();
+        VersionSelector preferredSelector = constraint.getPreferredSelector(versionSelectorScheme);
+        if (preferredSelector.isDynamic()) {
+            dynamicRevisionResolver.resolve(dependency, preferredSelector, result);
         } else {
-            DefaultModuleComponentIdentifier id = new DefaultModuleComponentIdentifier(requested.getGroup(), requested.getName(), requested.getVersion());
+            String version;
+            if (preferredSelector instanceof ExactVersionSelector) {
+                version = ((ExactVersionSelector) preferredSelector).getSelector();
```

Comment:
What is this for?
- The only `VersionSelector` where `isDynamic() == false` is `ExactVersionSelector`
- In this case,  `ExactVersionSelector.getSelector()` will always be equal to `ModuleVersionSelector.requested()`

I think you can safely leave the code as it was. Alternatively, we could add `VersionSelector.getSelector()` and use it always.

Reply:
This is already done in the next PR (adding `getSelector`)